File: leetcode/go/2016.maximum-difference-between-increasing-elements.go
func maximumDifference(nums []int) int {
    if len(nums) < 2 {
        return -1
    }
    s := make(stack, 0)
    ans := -1
    l := len(nums)
    s = s.Push(nums[l-1])
    for i:=l-2; i > -1 ; i-- {
        if nums[i] >= s.Peek() {
            s = s.Push(nums[i])
        } else {
            ans = Max(ans, s.Peek() - nums[i])
        }
    }
    return ans
}

func Max(a int, b int) int {
    if a > b {return a}
    return b
}

type stack []int

func (s stack) Peek() int {
    return s[len(s)-1]
}

func (s stack) Push(v int) stack {
    return append(s, v)
}

func (s stack) Pop() (stack, int) {
    l := len(s)
    return  s[:l-1], s[l-1]
}

File: leetcode/rust-leetcode/src/2016.maximum-difference-between-increasing-elements.rs
impl Solution {
    pub fn maximum_difference(nums: Vec<i32>) -> i32 {
        if nums.len() < 2 {
            return -1;
        }
        let mut ans = -1;
        let mut stack = vec![*nums.last().unwrap()];
        for i in (0..nums.len()-1).rev() {
            if nums[i] >= *stack.last().unwrap() {
                stack.push(nums[i]);
            } else {
                ans = ans.max(*stack.last().unwrap() - nums[i]);
            }
        }
        ans
    }
}

